compiler-plugin/src/main/kotlin/arrow/meta/quotes/ClassOrObject.kt

```diff
@@ -13,12 +13,111 @@ import org.jetbrains.kotlin.psi.psiUtil.getValueParameters
 import org.jetbrains.kotlin.psi.psiUtil.modalityModifierType
 import org.jetbrains.kotlin.psi.psiUtil.visibilityModifierType
 
+/**
+ * [classOrObject] is a function that intercepts all [KtClass] elements that [match]
+ * allowing a [Transform] to change the intercepted AST tree before compilation.
+ *
+ * An extension function of [Meta] and inheriting from [ExtensionPhase], [classOrObject] was designed to to feed in
+ * virtually any kind of [KtClass] predicate, followed by mapping a function that takes the desired [Scope] of our
+ * [KtClass] to change whatever PSI elements desired.
+ *
+ * For example, the [LensPlugin] and the [HigherKindPlugin] favors enabling ad-hoc polymorphism and more for the
+ * compiler by taking advantage of typeclass checks.
+ *
+ * In the [LensPlugin], we check to see if a type is a product type.  To add context, the cardinality of a product
+ * type is the product of their cardinalities:
+ *
+ * ```
+ *      |(a, b)| = |a| x |b|
+ * ```
+ *
+ * In particular, the [LensPlugin] is passing a function to whatever class or object type is a product type:
+ *
+ * ```
+ *  val Meta.lenses: Plugin
+ *    get() =
+ *      "lenses" {
+ *        meta {
+ *          classOrObject(::isProductOrType) { c -> ... }
+ *        }
+ *      }
+ * ```
+ *
+ * Where the function type for [::isProductType] is written as:
+ *
+ * ```
+ * fun isProductType(ktClass: KtClass): Boolean =
+ *      ktClass.isData() &&
+ *        ktClass.primaryConstructorParameters.isNotEmpty() &&
+ *        ktClass.primaryConstructorParameters.all { !it.isMutable } &&
+ *        ktClass.typeParameters.isEmpty()
+ * ```
+ *
+ * For a [KtClass] to be a product type, we check to see if the [KtClass] is an ADT, and that the cardinality of
+ * the constructor parameters is greater than one and non-mutable. For the [LensPlugin], we've just demonstrated how we
+ * can check for the product type of a [KtClass] by access the properties made available by the Kotlin PSI change.
```

Comment:
@nomisRev I thought it might be cool to include an example from arrow-meta to showcase FP capabilities we're enabling in the compiler on top of the basic example I give below. Sometimes not including more complicated examples limits the imagination of those reading documentation

Reply:
Yes, that's great especially since it'll also generate the website documentation!

We also include playground example snippets in Arrow, and hopefully with Arrow Meta IDE plugins we can give them support for playgrounds in the IDE :D

Reply:
Oh yeah, that was another thing too - so I have some executable code for a smaller example, the question I meant to ask is "should we have executable playground code for this example?" I wanted to reduce the code to focus on the bits that mattered most, but it's not totally consistent with the `runAnk` task allowing Kotlin to compile and execute in the simple example plugin.